pkg/gateway/gateway.go

```diff
@@ -0,0 +1,226 @@
+	return func(gw *Gateway) error {
+		var err error
+		sdk, err := fabsdk.New(config)
+
+		if err != nil {
+			return err
+		}
+
+		gw.sdk = sdk
+
+		configBackend, err := config()
+		if err != nil {
+			return err
+		}
+		if len(configBackend) != 1 {
+			return errors.New("invalid config file")
+		}
+
+		cfg := configBackend[0]
+		gw.cfg = cfg
+
+		value, ok := cfg.Lookup("client.organization")
+		if !ok {
+			return errors.New("No client organization defined in the config")
+		}
+		gw.org = value.(string)
+
+		return nil
+	}
+}
+
+// WithSDK configures the gateway with the configuration from an existing FabricSDK instance
+func WithSDK(sdk *fabsdk.FabricSDK) ConfigOption {
+	return func(gw *Gateway) error {
+		gw.sdk = sdk
+
+		cfg, err := sdk.Config()
+
+		if err != nil {
+			return errors.Wrap(err, "Unable to access SDK configuration")
+		}
+
+		value, ok := cfg.Lookup("client.organization")
+		if !ok {
+			return errors.New("No client organization defined in the config")
+		}
+		gw.org = value.(string)
+
+		return nil
+	}
+}
+
+// WithIdentity is an optional argument to the Connect method which specifies
+// the identity that is to be used to connect to the network.
+// All operations under this gateway connection will be performed using this identity.
+func WithIdentity(wallet Wallet, label string) IdentityOption {
+	return func(gw *Gateway) error {
+		mspClient, err := msp.New(gw.getSDK().Context(), msp.WithOrg(gw.getOrg()))
+		if err != nil {
+			return err
+		}
+
+		creds, err := wallet.Get(label)
+		if err != nil {
+			return err
+		}
+
+		var identity mspProvider.SigningIdentity
+		switch v := creds.(type) {
+		case *X509Identity:
+			identity, err = mspClient.CreateSigningIdentity(mspProvider.WithCert([]byte(v.Cert())), mspProvider.WithPrivateKey([]byte(v.Key())))
+			if err != nil {
+				return err
+			}
+		}
+
+		gw.options.Identity = identity
+		return nil
+	}
+}
+
+// WithUser is an optional argument to the Connect method which specifies
+// the identity that is to be used to connect to the network.
+// All operations under this gateway connection will be performed using this identity.
+func WithUser(user string) IdentityOption {
+	return func(gw *Gateway) error {
+		gw.options.User = user
+		return nil
+	}
+}
+
+// WithCommitHandler is an optional argument to the Connect method which
+// allows an alternative commit handler to be specified. The commit handler defines how
+// client code should wait to receive commit events from peers following submit of a transaction.
+// Currently unimplemented.
+func WithCommitHandler(handler CommitHandlerFactory) Option {
+	return func(gw *Gateway) error {
+		gw.options.CommitHandler = handler
+		return nil
+	}
+}
+
+// WithDiscovery is an optional argument to the Connect method which
+// enables or disables service discovery for all transaction submissions for this gateway.
+func WithDiscovery(discovery bool) Option {
+	return func(gw *Gateway) error {
+		gw.options.Discovery = discovery
```

Comment:
I guess this bool will be used in a future PR?

Reply:
Correct, it will done under this story:
https://jira.hyperledger.org/browse/FABG-941
This PR delivers the framework